test/memory_coloring_test.cpp

```diff
@@ -3374,7 +3374,7 @@ TEST_CASE(rnn_dom)
     m.add_instruction(pass_op{}, moutput, mx250, mx249, mx248);
 
     run_pass(m);
-    CHECK(m.get_parameter_shape("scratch").bytes() == 1600);
+    CHECK(m.get_parameter_shape("scratch").bytes() == 1824); // Optimal is 1600
```

Comment:
If 1600 is optimal, why are we now different? Is this a regression or needed due to the additional alignment/padding we've added?

Reply:
Graph coloring is NP-complete so it isnt always possible to get the best result. It depends on the order we process the nodes. This case uses more memory but other cases we use less memory.

Reply:
I don't think ordering of how the nodes are processed should matter here. You can reduce this into a decision problem + binary search to get the min amount of colors but I think the trade may be space and additional complexity.

Maybe as a later improvement but we can reduce the amount of colors if we use a SAT solver to determine the min amount of colors before finding the configuration of those colors. 

Like there's libraries like http://minisat.se/, that are under MIT license I've used to quickly find if the number of colors chosen there exists a solution, then again we'd need to represent things a certain way for the solver to work, and then use something like a binary search until we converged. 

I've got code for this I can dig up if this is something you want to explore further.

Reply:
> I don't think ordering of how the nodes are processed should matter here.

Since we are using a greedy algorithm for coloring the order does matter. From Wikipedia's [Greedy Coloring](https://en.wikipedia.org/wiki/Greedy_coloring#Choice_of_ordering):

> Different orderings of the vertices of a graph may cause the greedy coloring to use different numbers of colors, ranging from the optimal number of colors to, in some cases, a number of colors that is proportional to the number of vertices in the graph. For instance, a [crown graph](https://en.wikipedia.org/wiki/Crown_graph) (a graph formed from two [disjoint sets](https://en.wikipedia.org/wiki/Disjoint_sets) of n/2 vertices {a1, a2, ...} and {b1, b2, ...} by connecting ai to bj whenever i ≠ j) can be a particularly bad case for greedy coloring. With the vertex ordering a1, b1, a2, b2, ..., a greedy coloring will use n/2 colors, one color for each pair (ai, bi). However, the optimal number of colors for this graph is two, one color for the vertices ai and another for the vertices bi.[[5]](https://en.wikipedia.org/wiki/Greedy_coloring#cite_note-5) There also exist graphs such that with high probability a randomly chosen vertex ordering leads to a number of colors much larger than the minimum.[[6]](https://en.wikipedia.org/wiki/Greedy_coloring#cite_note-6) Therefore, it is of some importance in greedy coloring to choose the vertex ordering carefully.

There are some adaptive orderings that could improve on what we are currently doing.

Reply:
> if we use a SAT solver to determine the min amount of colors before finding the configuration of those colors.

That can tell use the minimum memory possible, but can it tell us the start and end of each of the segments as well?